algopy_testing/tests/primitives/test_biguint.py

```diff
@@ -0,0 +1,701 @@
+_too_big512_error = re.escape(f"expected value <= {MAX_UINT512}")
+
+_underflow_error = "- underflows"
+
+_avm_overflow_error = "math attempted on large byte-array"
+_avm_underflow_error = "byte math would have negative result"
+_avm_zero_division_error = "division by zero"
+_avm_zero_mod_error = "modulo by zero"
+
+
+def _int_to_bytes(x: int) -> bytes:
+    return x.to_bytes((x.bit_length() + 7) // 8, "big")
+
+
+def _object_as_biguint(x: object) -> BigUInt | object:
+    return BigUInt.from_bytes(x) if (isinstance(x, bytes)) else x
+
+
+def _get_avm_result(
+    get_avm_result: AVMInvoker, method: str, a: int, b: int | UInt64
+) -> BigUInt | object:
+    a_bytes = _int_to_bytes(a)
+    if isinstance(b, UInt64):
+        avm_result = get_avm_result(method, a=a_bytes, b=b.value)
+    else:
+        b_bytes = _int_to_bytes(b)
+        avm_result = get_avm_result(method, a=a_bytes, b=b_bytes)
+    return _object_as_biguint(avm_result)
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        0,
+        1,
+        MAX_UINT512,
+    ],
+)
+def test_biguint_unary_plus(value: int) -> None:
+    # just test the implementation as there is no AVM equivalent
+    assert +BigUInt(value) == value
+
+
+@pytest.mark.parametrize(
+    "value",
+    [
+        0,
+        1,
+        MAX_UINT512,
+    ],
+)
+def test_biguint_bool(value: int) -> None:
+    # just test the implementation as there is no AVM equivalent
+    assert bool(BigUInt(value)) == bool(value)
+
+
+@pytest.mark.parametrize(
+    ("a", "b"),
+    [
+        (0, 0),
+        (0, 1),
+        (0, MAX_UINT512),
+        (1, 0),
+        (1, 1),
+        (1, MAX_UINT512),
+        (MAX_UINT512, MAX_UINT512),
+    ],
+)
+@pytest.mark.parametrize(
+    "op_name",
+    [
+        "eq",
+        "ne",
+        "lt",
+        "le",
+        "gt",
+        "ge",
+    ],
+)
+def test_biguint_comparison(get_avm_result: AVMInvoker, op_name: str, a: int, b: int) -> None:
+    avm_result = _get_avm_result(get_avm_result, f"verify_biguint_{op_name}", a, b)
+    op = getattr(operator, op_name)
+
+    assert avm_result == op(BigUInt(a), BigUInt(b))
+    assert avm_result == op(BigUInt(a), b)
+    assert avm_result == op(a, BigUInt(b))
+
+
+@pytest.mark.parametrize(
+    ("a", "b"),
+    [
+        (0, UInt64(0)),
+        (0, UInt64(1)),
+        (0, UInt64(MAX_UINT64)),
+        (1, UInt64(0)),
+        (1, UInt64(1)),
+        (1, UInt64(MAX_UINT64)),
+        (MAX_UINT512, UInt64(MAX_UINT64)),
+        (MAX_UINT64, UInt64(MAX_UINT64)),
+    ],
+)
+@pytest.mark.parametrize(
+    "op_name",
+    [
+        "eq",
+        "ne",
+        "lt",
+        "le",
+        "gt",
+        "ge",
+    ],
+)
+def test_biguint_comparison_uint64(
+    get_avm_result: AVMInvoker, op_name: str, a: int, b: UInt64
+) -> None:
+    avm_result = _get_avm_result(get_avm_result, f"verify_biguint_{op_name}_uint64", a, b)
+    op = getattr(operator, op_name)
+    assert avm_result == op(BigUInt(a), b)
+    if a < MAX_UINT64:
```

Comment:
MAX_UINT64 is inclusive
```suggestion
    if a <= MAX_UINT64:
```